### ISSUE TITLE
CUDA Install error

### ISSUE THREAD

**Issue by Martin11847** [Nebula-simulator/Nebula#4]
When trying to follow the installation procedure, I met the following error :
`#error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.`
I use nvcc version 11.5

**Comment by jbelissard** [comment]
> When trying to follow the installation procedure, I met the following error : `#error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.` I use nvcc version 11.5

Hi Martin !
You need to add these lines in this file  : nebula/3rdparty/cub/cub/util_namespace.cuh
`#ifndef CUB_NS_QUALIFIER`
`#define CUB_NS_QUALIFIER ::cub`
`#endif`

It worked for me, so I hope it will work for you too.

**Comment by Martin11847** [comment]
It worked for me too, thanks a lot !

**Comment by lvkessel** [comment]
It seems that this is a problem introduced with recent versions of the CUDA toolkit. It also seems that this can be fixed by updating the cub library. I have done that in a new branch (update-cub), but sadly I do not have the ability to test that at the moment. I am keeping this issue open until we have a confirmed working fix.

**Comment by uowei** [comment]
> > When trying to follow the installation procedure, I met the following error : `#error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.` I use nvcc version 11.5
> 
> Hi Martin ! You need to add these lines in this file : nebula/3rdparty/cub/cub/util_namespace.cuh `#ifndef CUB_NS_QUALIFIER` `#define CUB_NS_QUALIFIER ::cub` `#endif`
> 
> It worked for me, so I hope it will work for you too.

@Martin11847 @jbelissard @lvkessel 
I added the lines you mentioned, but it seems that another error has occurred. 

```
/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh(778): error: more than one instance of overloaded function "cub::Debug" matches the argument list:
            function "cub::CUB_200200_600_610_700_750_800_860_NS::Debug(cudaError_t, const char *, int)" (declared at line 184 of /usr/local/cuda-12.2/targets/x86_64-linux/include/cub/util_debug.cuh)
            function "cub::Debug(cudaError_t, const char *, int)" (declared at line 68 of /mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/block/specializations/../../block/../util_debug.cuh)
            argument types are: (cudaError_t, const char [109], int)
              if (cub::Debug((cudaError_t) (error = spmv_params.t_vector_x.UnbindTexture()), "/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh", 778)) break;
                  ^

/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh(806): error: more than one instance of overloaded function "cub::Debug" matches the argument list:
            function "cub::CUB_200200_600_610_700_750_800_860_NS::Debug(cudaError_t, const char *, int)" (declared at line 184 of /usr/local/cuda-12.2/targets/x86_64-linux/include/cub/util_debug.cuh)
            function "cub::Debug(cudaError_t, const char *, int)" (declared at line 68 of /mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/block/specializations/../../block/../util_debug.cuh)
            argument types are: (cudaError_t, const char [109], int)
              if (cub::Debug((cudaError_t) (error = PtxVersion(ptx_version)), "/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh", 806)) break;
                  ^

/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh(815): error: more than one instance of overloaded function "cub::Debug" matches the argument list:
            function "cub::CUB_200200_600_610_700_750_800_860_NS::Debug(cudaError_t, const char *, int)" (declared at line 184 of /usr/local/cuda-12.2/targets/x86_64-linux/include/cub/util_debug.cuh)
            function "cub::Debug(cudaError_t, const char *, int)" (declared at line 68 of /mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/block/specializations/../../block/../util_debug.cuh)
            argument types are: (cudaError_t, const char [109], int)
              if (cub::Debug((cudaError_t) (error = Dispatch( d_temp_storage, temp_storage_bytes, spmv_params, stream, debug_synchronous, DeviceSpmv1ColKernel<PtxSpmvPolicyT, ValueT, OffsetT>, DeviceSpmvSearchKernel<PtxSpmvPolicyT, OffsetT, CoordinateT, SpmvParamsT>, DeviceSpmvKernel<PtxSpmvPolicyT, ScanTileStateT, ValueT, OffsetT, CoordinateT, false, false>, DeviceSegmentFixupKernel<PtxSegmentFixupPolicy, KeyValuePairT*, ValueT*, OffsetT, ScanTileStateT>, spmv_config, segment_fixup_config)), "/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh", 815)
                  ^

66 errors detected in the compilation of "/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/source/gpu_main.cu".
make[2]: *** [source/CMakeFiles/nebula_gpu.dir/build.make:77: source/CMakeFiles/nebula_gpu.dir/gpu_main.cu.o] Error 2
make[1]: *** [CMakeFiles/Makefile2:138: source/CMakeFiles/nebula_gpu.dir/all] Error 2
make: *** [Makefile:91: all] Error 2
```



Could you please advise on how to resolve this issue?
thanks!

**Comment by nowinder** [comment]
> > > When trying to follow the installation procedure, I met the following error : `#error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.` I use nvcc version 11.5
> > 
> > 
> > Hi Martin ! You need to add these lines in this file : nebula/3rdparty/cub/cub/util_namespace.cuh `#ifndef CUB_NS_QUALIFIER` `#define CUB_NS_QUALIFIER ::cub` `#endif`
> > It worked for me, so I hope it will work for you too.
> 
> @Martin11847 @jbelissard @lvkessel I added the lines you mentioned, but it seems that another error has occurred.
> 
> ```
> /mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh(778): error: more than one instance of overloaded function "cub::Debug" matches the argument list:
>             function "cub::CUB_200200_600_610_700_750_800_860_NS::Debug(cudaError_t, const char *, int)" (declared at line 184 of /usr/local/cuda-12.2/targets/x86_64-linux/include/cub/util_debug.cuh)
>             function "cub::Debug(cudaError_t, const char *, int)" (declared at line 68 of /mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/block/specializations/../../block/../util_debug.cuh)
>             argument types are: (cudaError_t, const char [109], int)
>               if (cub::Debug((cudaError_t) (error = spmv_params.t_vector_x.UnbindTexture()), "/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh", 778)) break;
>                   ^
> 
> /mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh(806): error: more than one instance of overloaded function "cub::Debug" matches the argument list:
>             function "cub::CUB_200200_600_610_700_750_800_860_NS::Debug(cudaError_t, const char *, int)" (declared at line 184 of /usr/local/cuda-12.2/targets/x86_64-linux/include/cub/util_debug.cuh)
>             function "cub::Debug(cudaError_t, const char *, int)" (declared at line 68 of /mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/block/specializations/../../block/../util_debug.cuh)
>             argument types are: (cudaError_t, const char [109], int)
>               if (cub::Debug((cudaError_t) (error = PtxVersion(ptx_version)), "/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh", 806)) break;
>                   ^
> 
> /mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh(815): error: more than one instance of overloaded function "cub::Debug" matches the argument list:
>             function "cub::CUB_200200_600_610_700_750_800_860_NS::Debug(cudaError_t, const char *, int)" (declared at line 184 of /usr/local/cuda-12.2/targets/x86_64-linux/include/cub/util_debug.cuh)
>             function "cub::Debug(cudaError_t, const char *, int)" (declared at line 68 of /mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/block/specializations/../../block/../util_debug.cuh)
>             argument types are: (cudaError_t, const char [109], int)
>               if (cub::Debug((cudaError_t) (error = Dispatch( d_temp_storage, temp_storage_bytes, spmv_params, stream, debug_synchronous, DeviceSpmv1ColKernel<PtxSpmvPolicyT, ValueT, OffsetT>, DeviceSpmvSearchKernel<PtxSpmvPolicyT, OffsetT, CoordinateT, SpmvParamsT>, DeviceSpmvKernel<PtxSpmvPolicyT, ScanTileStateT, ValueT, OffsetT, CoordinateT, false, false>, DeviceSegmentFixupKernel<PtxSegmentFixupPolicy, KeyValuePairT*, ValueT*, OffsetT, ScanTileStateT>, spmv_config, segment_fixup_config)), "/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/3rdparty/cub/cub/device/dispatch/dispatch_spmv_orig.cuh", 815)
>                   ^
> 
> 66 errors detected in the compilation of "/mnt/c/Users/MediaCore_PC/Desktop/TSMC/elsepa/nebula/source/gpu_main.cu".
> make[2]: *** [source/CMakeFiles/nebula_gpu.dir/build.make:77: source/CMakeFiles/nebula_gpu.dir/gpu_main.cu.o] Error 2
> make[1]: *** [CMakeFiles/Makefile2:138: source/CMakeFiles/nebula_gpu.dir/all] Error 2
> make: *** [Makefile:91: all] Error 2
> ```
> 
> Could you please advise on how to resolve this issue? thanks!

hello uowei ! 
did you find the solution to your issue? When the error `#error CUB requires a definition of CUB_NS_QUALIFIER when CUB_NS_PREFIX/POSTFIX are defined.` reported and i add the three lines the similar error occured in my windows cuda12.5 env too.

**Comment by nowinder** [comment]
As the issue always seems to be from the cub, i replace the whole cub subdirectory under `3rdparty/cub/` with the cub directory corresponding to the cuda version which under the `ROOTDIR\NVIDIA GPU Computing Toolkit\CUDA\v12.5\include\`(for my windows cuda12.5 env), and the it works. :-)

**Comment by uowei** [comment]
@nowinder 
I redid the installation according to the instructions, and it worked without any issues. The only changes I made were to the versions of CMake and HDF5; my CMake version is 3.26.5, and HDF5 is 1.8.21. I suspect that there may have been conflicts between these versions, but I am not sure of the exact reason. However, the images I generated consistently have thin black lines on the far left, regardless of the position on any graphical representation.

![800_100_01](https://github.com/user-attachments/assets/ccd283d4-a381-4853-a372-549364b5acc4)

**Comment by lvkessel** [comment]
Closed by a50a8e8. Given that it's a real pain to support both modern and old versions of cuda at the same time, we are dropping support for cuda < 9.2 (even though the code would technically still work with that). Probably fine though, given that 9.2 is a pretty old version by now.
